kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h

```diff
@@ -0,0 +1,1234 @@
+                        empty = false;
+                        break;
+                    }
+                }
+
+                if(empty) {
+                    rA(k, k) = BaseType::mScaleFactor;
+                    rb[k] = 0.0;
+                }
+            }
+        }
+
+        #pragma omp parallel for firstprivate(system_size)
+        for (int k = 0; k < static_cast<int>(system_size); ++k) {
+            std::size_t col_begin = Arow_indices[k];
+            std::size_t col_end = Arow_indices[k+1];
+            const double k_factor = scaling_factors[k];
+            if (k_factor == 0.0) {
+                // Zero out the whole row, except the diagonal
+                for (std::size_t j = col_begin; j < col_end; ++j)
+                    if (static_cast<int>(Acol_indices[j]) != k )
+                        Avalues[j] = 0.0;
+
+                // Zero out the RHS
+                rb[k] = 0.0;
+            } else {
+                // Zero out the column which is associated with the zero'ed row
+                for (std::size_t j = col_begin; j < col_end; ++j)
+                    if(scaling_factors[ Acol_indices[j] ] == 0 )
+                        Avalues[j] = 0.0;
+            }
+        }
+    }
+    
+    /**
+     * @brief Applies the constraints with master-slave relation matrix (RHS only)
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rb The RHS vector
+     */
+    void ApplyRHSConstraints(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemVectorType& rb
+        ) override
+    {
+        KRATOS_TRY
+
+        if (rModelPart.MasterSlaveConstraints().size() != 0) {
+            // First we check if CONSTRAINT_SCALE_FACTOR is defined
+            if (mConstraintFactorConsidered != CONSTRAINT_FACTOR::CONSIDER_PRESCRIBED_CONSTRAINT_FACTOR) {
+                TSystemMatrixType A(BaseType::mEquationSystemSize, BaseType::mEquationSystemSize);
+                BaseType::ConstructMatrixStructure(pScheme, A, rModelPart);
+                this->BuildLHS(pScheme, rModelPart, A);
+                const double constraint_scale_factor = mConstraintFactorConsidered == CONSTRAINT_FACTOR::CONSIDER_NORM_DIAGONAL_CONSTRAINT_FACTOR ? this->GetDiagonalNorm(A) : this->GetDiagonalNorm(A);
+                mConstraintFactor = constraint_scale_factor;
+            }
+
+            // Check T has been computed
+            if (TSparseSpace::Size1(BaseType::mT) != BaseType::mSlaveIds.size() || TSparseSpace::Size2(BaseType::mT) != BaseType::mEquationSystemSize) {
+                BaseType::mT.resize(BaseType::mSlaveIds.size(), BaseType::mEquationSystemSize, false);
+                ConstructMasterSlaveConstraintsStructure(rModelPart);
+            }
+
+            // If not previously computed we compute
+            if (TSparseSpace::TwoNorm(BaseType::mT) < std::numeric_limits<double>::epsilon()) {
```

Comment:
it is expensive to compute the norm of T. Why do you need to do it this way??

Reply:
I can add a flag to know if the mT has been already computed

Reply:
I forgot about this, one sec

Reply:
Done